### PR TITLE
v3: sourceKey feature for hasMany doesn't work if where clause specified in include (#7141)

### PR DESCRIPTION
**Pull request by LoneWolfPR**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Have you added an entry under `Future` in the changelog?

_NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._

### Description of change

This fixes the issue mentioned in https://github.com/sequelize/sequelize/issues/7141. Previously `sourceKey` was only respected in a find from hasMany association if the include did not contain a where clause. This fixes that issue so that even with a where clause specified the `sourceKey` supplied by the user will be respected.

### CODE REVIEW

**Comment by LoneWolfPR** [comment]
Any chance this can get merged in sometime soon (assuming it meets all merge requirements)? We have a project that needs this pretty badly with multiple developers needing it.

File: lib/dialects/abstract/query-generator.js [hunk 1]
```diff
@@ -1612,7 +1607,6 @@ var QueryGenerator = {
     }
 
     query += ';';
-
```
Comment (janmeier) [inline lib/dialects/abstract/query-generator.js]: Please remove the whitespace changes (it makes the changes easier to cherry-pick onto master)

File: test/integration/associations/has-many.test.js [hunk 2]
```diff
@@ -1285,4 +1285,45 @@ describe(Support.getTestDialectTeaser('HasMany'), function() {
 
   });
 
+  describe('sourceKey with where clause in include', function() {
+    beforeEach(function() {
+      var User = this.sequelize.define('User',
+        { username: Sequelize.STRING, email: Sequelize.STRING },
```
Comment (janmeier) [inline test/integration/associations/has-many.test.js]: Could you change it to something that uses `field` (`email: { type: Sequelize.STRING, field: 'mail' }` or similar - So we can make sure this works (and continues to work) even when the field is called something other in the DB than in the model

Comment (janmeier) [inline test/integration/associations/has-many.test.js]: Username attribute and indexes are not needed for this test

File: test/integration/associations/has-many.test.js [hunk 3]
```diff
@@ -1285,4 +1285,45 @@ describe(Support.getTestDialectTeaser('HasMany'), function() {
...
+            include: [
+              {
+                model: Task,
+                where: {taskStatus: 'Active'}
+              }
+            ],
+            where: { username: 'John' }
+          }).then(function(user) {
+            expect(user).to.be.ok;
+            expect(user.Tasks.length).to.equal(1);
```
Comment (janmeier) [inline test/integration/associations/has-many.test.js]: Please test that we get the correct task as well

**Comment by LoneWolfPR** [comment]
@janmeier I've addressed all your requests. I believe it should be good to go now. Thanks for reviewing. Let me know if there are anymore issues.

**Comment by janmeier** [comment]
Cherry-picked in https://github.com/sequelize/sequelize/commit/7d3172dfe4a46aaecc1492c42cc86f996bd663d2

**Comment by janmeier** [comment]
Thanks for the adressing my feedback quickly @LoneWolfPR  :tada:

**Comment by LoneWolfPR** [comment]
@janmeier no problem. Glad I could help. What's the timeline for this going out to the npm?

**Comment by janmeier** [comment]
`+ sequelize@3.30.1`

**Comment by danconnell** [comment]
Thanks @LoneWolfPR. Just found this bug myself and was very happy to see it fixed already 😄

**Comment by derimagia** [comment]
I've been having an issue here where the field name is not the same as the model property name. So, if the 'field' is set and it's different the column name this isn't working.  

https://github.com/sequelize/sequelize/blob/3168d543f8d488f7c7db2d3d9d082762c288789a/lib/associations/has-many.js#L88-L98

Line 88 is using `this.target` and line 89 is using `this.source`, shouldn't they both be checking the source? 

Once `this.sourceKeyField`is fixed then `this.sourceKey` just needs to be replaced by it on lines 94-95. I'd be happy yo put a pull request if need be but wanted to check with someone who knows this more to correct me if I'm missing something.

**Pull request merged**